### PR TITLE
Add support for `tool.uv` into distribution building

### PR DESCRIPTION
**Pull request by konstin**
With the change, we remove the special casing of workspace dependencies and resolve `tool.uv` for all git and directory distributions. This gives us support for non-editable workspace dependencies and path dependencies in other workspaces. It removes a lot of special casing around workspaces. These changes are the groundwork for supporting `tool.uv` with dynamic metadata.

The basis for this change is moving `Requirement` from `distribution-types` to `pypi-types` and the lowering logic from `uv-requirements` to `uv-distribution`. This changes should be split out in separate PRs.

I've included an example workspace `albatross-root-workspace2` where `bird-feeder` depends on `a` from another workspace `ab`. There's a bunch of failing tests and regressed error messages that still need fixing. It does fix the audited package count for the workspace tests.

### CODE REVIEW

**Comment by charliermarsh** [comment]
Can `ArchiveMetadata` just be removed entirely? Do we need both `ArchiveMetadata` and `ArchiveMetadataLowered`?

File: crates/uv-distribution/src/lib.rs [hunk 1]
```diff
@@ -35,3 +49,133 @@ impl From<Metadata23> for ArchiveMetadata {
...
+pub enum MetadataLoweringError {
+    #[error(transparent)]
+    Workspace(#[from] WorkspaceError),
+    #[error(transparent)]
+    Lowering(#[from] LoweringError),
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "kebab-case")]
+pub struct Metadata23Lowered {
```
Comment (charliermarsh) [inline crates/uv-distribution/src/lib.rs:63]: Could we just call this `Metadata` or something, differentiated from `Metadata23` from the spec? Or use `pypi_types::Metadata23` everywhere?

Comment (konstin) [reply crates/uv-distribution/src/lib.rs:63]: Absolutely these names are only for me to be easier to work with and suppose to change before merging, ideally we'd even remove some type

File: crates/uv-requirements/src/specification.rs [hunk 2]
```diff
@@ -564,17 +279,274 @@ impl RequirementsSpecification {
...
-            preview,
-        )
-        .await
+        Self::from_sources(requirements, &[], &[], client_builder).await
+    }
+}
+
+// TODO(konsti): Hook this into source tree handling.
+/*
+#[cfg(test)]
```
Comment (konstin) [inline crates/uv-requirements/src/specification.rs:289]: These were intended as unit tests for the lowering, but the api that i used to the reading went away so now we need to hook this somewhere different

File: crates/uv-distribution/src/source/mod.rs [hunk 3]
```diff
@@ -1137,7 +1156,10 @@ impl<'a, T: BuildContext> SourceDistributionBuilder<'a, T> {
         {
             if let Some(metadata) = read_cached_metadata(&metadata_entry).await? {
                 debug!("Using cached metadata for: {source}");
-                return Ok(ArchiveMetadata::from(metadata));
+                return Ok(ArchiveMetadataLowered::from(
+                    Metadata23Lowered::from_tool_uv(metadata, fetch.path(), self.preview_mode)
```
Comment (charliermarsh) [inline crates/uv-distribution/src/source/mod.rs:1160]: @konstin - Do you think we should be caching the lowered or non-lowered metadata here? What are the implications of caching the non-lowered metadata, then lowering it after?

**Comment by charliermarsh** [comment]
@konstin - This is passing tests on Linux and macOS now, one Windows-specific failure.

File: crates/uv-distribution/src/lib.rs [hunk 4]
```diff
@@ -15,20 +25,118 @@ mod error;
...
+                .map(pypi_types::Requirement::from)
+                .collect(),
+            requires_python: metadata.requires_python,
+            provides_extras: metadata.provides_extras,
+        }
+    }
+
+    /// Lower by considering `tool.uv` in `pyproject.toml` if present, used for Git and directory
+    /// dependencies.
+    pub async fn from_workspace(
```
Comment (charliermarsh) [inline crates/uv-distribution/src/lib.rs:72]: Kind of wish none of this were async. Wondering if it shouldn't be given that it's all fs operations.

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
The member globbing isn't working properly on Windows.

**Comment by charliermarsh** [comment]
I think I fixed it but need to rebase.

**Comment by charliermarsh** [comment]
@konstin - I got this into a mergable state, but I removed `ab`, `albatross-root-workspace2`, and the commented-out tests in `specification.rs`. Can you restore any of those that are relevant in a separate PR?

I also made workspace discovery preview-only to make this a bit safer.

File: crates/uv-distribution/src/lib.rs (hunk 4 above)
Comment (konstin) [reply crates/uv-distribution/src/lib.rs:71]: Like i said in the initial PR, those are only async for consistency, we can totally make them sync

**Comment by konstin** [comment]
Restored the commented out tests, wrote a follow-up in https://github.com/astral-sh/uv/issues/3943

**Comment by charliermarsh** [comment]
Thanks!

**Pull request merged**